hbase-asyncfs/src/main/java/org/apache/hadoop/hbase/io/asyncfs/monitor/StreamSlowMonitor.java

```diff
@@ -0,0 +1,176 @@
+
+  /**
+   * Configure for the packet flush speed.
+   */
+  private static final String DATANODE_SLOW_PACKET_FLUSH_SPEED_KEY =
+    "hbase.regionserver.async.wal.datanode.slow.packet.flush.speed.kbs";
+  private static final double DEFAULT_DATANODE_SLOW_PACKET_FLUSH_SPEED = 0.1;
+
+  private final String name;
+  // this is a map of datanodeInfo->queued slow PacketAckData
+  private final Map<DatanodeInfo, Deque<PacketAckData>> datanodeSlowDataQueue =
+    new ConcurrentHashMap<>();
+  private final ExcludeDatanodeManager excludeDatanodeManager;
+
+  private int minSlowDetectCount;
+  private long slowDataTtl;
+  private long slowPacketAckMillis;
+  private double slowPacketAckSpeed;
+
+  public StreamSlowMonitor(Configuration conf, String name,
+      ExcludeDatanodeManager excludeDatanodeManager) {
+    setConf(conf);
+    this.name = name;
+    this.excludeDatanodeManager = excludeDatanodeManager;
+    LOG.info("New stream slow monitor {}", this.name);
+  }
+
+  public static StreamSlowMonitor create(Configuration conf, String name) {
+    return new StreamSlowMonitor(conf, name, new ExcludeDatanodeManager(conf));
+  }
+
+  /**
+   * Check if the packet process time shows that the relevant datanode is a slow node.
+   * @param datanodeInfo the datanode that processed the packet
+   * @param packetDataLen the data length of the packet
+   * @param processTime the process time of the packet on the datanode
+   * @param lastAckTimestamp the last acked timestamp of the packet on another datanode
+   * @param unfinished if the packet is unfinished flushed to the datanode replicas
+   */
+  public void checkProcessTimeAndSpeed(DatanodeInfo datanodeInfo, long packetDataLen,
+      long processTime, long lastAckTimestamp, int unfinished) {
+    long current = EnvironmentEdgeManager.currentTime();
+    boolean slow = processTime > slowPacketAckMillis ||
+        (packetDataLen > 100 && (double) packetDataLen / processTime < slowPacketAckSpeed);
+    if (slow) {
+      // check if large diff ack timestamp between replicas
+      if ((lastAckTimestamp > 0 && current - lastAckTimestamp > slowPacketAckMillis / 2) || (
+          lastAckTimestamp <= 0 && unfinished == 0)) {
+        LOG.info("Slow datanode: {}, data length={}, duration={}ms, unfinishedReplicas={}, "
+            + "lastAckTimestamp={}, monitor name: {}", datanodeInfo, packetDataLen, processTime,
+          unfinished, lastAckTimestamp, this.name);
+        if (addSlowAckData(datanodeInfo, packetDataLen, processTime)) {
+          excludeDatanodeManager.tryAddExcludeDN(datanodeInfo, "slow packet ack");
+        }
+      }
+    }
+  }
+
+  @Override
+  public void onConfigurationChange(Configuration conf) {
+    setConf(conf);
+  }
+
+  private boolean addSlowAckData(DatanodeInfo datanodeInfo, long dataLength, long processTime) {
+    Deque<PacketAckData> slowDNQueue = datanodeSlowDataQueue.computeIfAbsent(datanodeInfo,
```

Comment:
When will we remove entries from this Map and the Deque? So once a datanode is marked as slow, it will always be marked as slow? And finally the map will contain all the datanodes in the cluster?

Reply:
The old data in the Deque is lazily cleared which is triggered by the newly coming slow data. So if a datanode is detected slow once, then there is at least one slow data for it in the memory...
And currently nowhere to clear the Map,  the Map will increased to contain all the datanodes in the worst case, though only slow packet ack triggers the datanode be recorded in the Map.
We can also use the guava cache to replace the Map here, and set a proper expireAfterWrite time for it.
What do you think? @Apache9